js/util/pipeline-fixed.js

```diff
@@ -0,0 +1,385 @@
+/**
+ * Copyright (C) 2013 Regents of the University of California.
+ * @author: Chavoosh Ghasemi <chghasemi@cs.arizona.edu>
+ * @author: Jeff Thompson <jefft0@remap.ucla.edu>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * A copy of the GNU Lesser General Public License is in the file COPYING.
+ */
+
+/** @ignore */
+var Interest = require('../interest.js').Interest; /** @ignore */
+var Blob = require('./blob.js').Blob; /** @ignore */
+var KeyChain = require('../security/key-chain.js').KeyChain; /** @ignore */
+var NdnCommon = require('./ndn-common.js').NdnCommon;
+var DataFetcher = require('./data-fetcher.js').DataFetcher;
+var LOG = require('../log.js').Log.LOG;
+
+
+/**
+ * Retrieve the segments of solicited data by keeping a fixed-size window of N
+ * of on fly Interests at any given time.
+ *
+ * To handle timeout and nack we use DataFetcher class which upon facing
+ * timeout or nack will try to resolve the corresponded segment by retransmitting
+ * the Interest a few times.
+ *
+ * PipelineFixed assumes that the data is named /<prefix>/<version>/<segment>,
+ * where:
+ * - <prefix> is the specified name prefix,
+ * - <version> is an unknown version that needs to be discovered, and
+ * - <segment> is a segment number. 
+ * Note: The number of segments is unknown and is controlled by the
+ *       `FinalBlockId` field in the first retrieved Data packet. So, the
+ *       very first Data packet MUST contain the `FinalBlockId` field.
+ *
+ * The following logic is implemented in PipelineFixed:
+ *
+ * 1. Express the first Interest to discover the version:
+ *
+ *    >> Interest: /<prefix>?MustBeFresh=true
+ *
+ * 2. Infer the latest version of the Data: <version> = Data.getName().get(-2)
+ *
+ * 3. If the segment number in the retrieved packet == 0, go to step 5, while the
```

Comment:
"go to step 5" is incorrect: there would be no Interests for segments [1, M-1)

Reply:
Are you talking about a case where the content only has 1 chunk (i.e., chunk zero)??? If so, that is just a corner case.